### PR TITLE
Closes #5090: Wire up GeckoView action delegate for BrowserActions

### LINKED ISSUE #5090: Wire up GeckoView action delegate for BrowserActions
This is a follow-up for https://github.com/mozilla-mobile/android-components/issues/4469 now that the GV API has landed.

### PR DESCRIPTION
**Pull request by csadilek**
This follows the same pattern as content/background messaging, with the only difference that we have to be able to register action handlers for each extension on every session. The latter is the reason I moved the session-specific action out of the `EngineObserver` into our web extension module. Ultimately, we want to get rid of`EngineObserver` anyway once we fully migrated to browser-state.

I am also going to add a browser action to our sample-browser extension, but I will do this as part of #4791, as this PR is already pretty big.

We will get new GV API for installing extensions i.e different calls for built-in and third-party extensions. For now I've also introduced a `supportActions` boolean, similar to our `allowContentMessaging` to be able to configure per extension if we need to hook up action delegates. We really don't need that for our current extensions, so it makes sense not to register all these handlers.

### CODE REVIEW

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt [hunk 1]
```diff
@@ -142,6 +151,64 @@ class GeckoWebExtension(
...
+
+        nativeExtension.setActionDelegate(actionDelegate)
+    }
+
+    /**
+     * See [WebExtension.registerActionHandler].
+     */
+    override fun registerActionHandler(session: EngineSession, actionHandler: ActionHandler) {
+        if (!supportActions) {
+            logger.error("Attempt to register action handler on session but browser and page " +
```
Comment (Amejia481) [inline components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt]: Nice 👍

Comment (csadilek) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt]: Hehe, yers, I forgot that before :). I just want to turn this off by default for built-in extensions, we can turn it on for third-party extensions, once we get the different methods.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/WebExtensionToolbarAction.kt [hunk 2]
```diff
@@ -38,16 +45,32 @@ open class WebExtensionToolbarAction(
 
         rootView.setBackgroundResource(backgroundResource)
         padding?.let { rootView.setPadding(it) }
+
+        rootView.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
```
Comment (Amejia481) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/WebExtensionToolbarAction.kt]: Nice ✨ 
Similar to the [pattern used in](https://github.com/mozilla-mobile/android-components/blob/6fef6179e1f4bfe59359b86efa7035116b32237a/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt#L179), it looks like this is a common need, maybe in the future we can have an utility extension for loading an image  async for everybody to use :)

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarFeatureTest.kt [hunk 3]
```diff
@@ -94,22 +94,20 @@ class WebExtensionToolbarFeatureTest {
 
         val browserAction = BrowserAction(
             title = "title",
-            icon = mock(),
+            loadIcon = { mock() },
```
Comment (Amejia481) [inline components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarFeatureTest.kt]: nit: Does worth it adding extra test coverage for the icon loading section and the `iconJob?.cancel()`  ?

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt [hunk 4]
```diff
@@ -4,53 +4,81 @@
...
 import org.junit.runner.RunWith
 import org.mockito.Mockito.verify
 
 @RunWith(AndroidJUnit4::class)
 class WebExtensionToolbarTest {
+    private val testDispatcher = TestCoroutineDispatcher()
+
+    @Before
+    @ExperimentalCoroutinesApi
+    fun setUp() {
```
Comment (Amejia481) [inline components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt]: nit: I think we could replace this block with [MainCoroutineRule](https://github.com/mozilla-mobile/android-components/blob/9b8e27603efea171c8f3764de9a25cc3f1e84ec9/components/support/test/src/main/java/mozilla/components/support/test/rule/MainCoroutineRule.kt#L28)

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt [hunk 5]
```diff
@@ -4,53 +4,81 @@
...
 
         val action = WebExtensionToolbarAction(browserAction) {}
-
         action.bind(view)
+        action.iconJob?.joinBlocking()
+        testDispatcher.advanceUntilIdle()
+
+        val iconCaptor = argumentCaptor<BitmapDrawable>()
+        verify(imageView).setImageDrawable(iconCaptor.capture())
+        assertEquals(icon, iconCaptor.value.bitmap)
```
Comment (Amejia481) [inline components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt]: nit: I'm not sure if it worths adding extra test coverage for the `iconJob?.cancel()`, what do you think?

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt [hunk 6]
```diff
@@ -63,10 +89,60 @@ object WebExtensionSupport {
...
+                    }
             }
         })
     }
 
+    /**
+     * Observes the provided store to register session-specific [ActionHandler]s
+     * for all installed extensions on newly added sessions.
+     */
+    private fun registerActionHandlersForNewSessions(store: BrowserStore) {
```
Comment (Amejia481) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: Could you help me to clarify `newly added sessions` term mean? Does that mean session that are just linked to its respective `engineSession`, right?

Comment (csadilek) [reply components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: Sure, we're already capturing the existing sessions when the extension is installed. So, for every session added after (once it's engine session is created / linked) we have to do this.

Comment (Amejia481) [reply components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: Thanks 👍

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt [hunk 7]
```diff
@@ -63,10 +89,60 @@ object WebExtensionSupport {
...
+                        action: BrowserAction
+                    ) {
+                        store.dispatch(WebExtensionAction.UpdateBrowserAction(webExtension.id, action))
+                    }
+                })
+
+                // Register action handler for all existing engine sessions on the new extension
+                store.state.tabs
+                    .forEach {
+                        val session = it.engineState.engineSession
```
Comment (psymoon) [inline components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: nit:
```						
it.engineState.engineSession?.let { session ->
	registerSessionActionHandler(webExtension, session, SessionActionHandler(store, it.id))
}
```

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt (hunk 4 above)
Comment (csadilek) [reply components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt]: Ah, nice! Updated.

File: components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt (hunk 5 above)
Comment (csadilek) [reply components/feature/toolbar/src/test/java/mozilla/components/feature/toolbar/WebExtensionToolbarTest.kt]: Yeah, might as well. Added a test now to make sure we cancel loading the icon.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/BrowserAction.kt [hunk 8]
```diff
@@ -21,8 +20,7 @@ import android.graphics.drawable.Drawable
 data class BrowserAction(
     val title: String,
     val enabled: Boolean,
-    val icon: Drawable,
-    val uri: String,
+    val loadIcon: suspend (Int) -> Bitmap?,
```
Comment (psymoon) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/BrowserAction.kt]: Is this related to AMO add-on icon?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt [hunk 9]
```diff
@@ -81,6 +84,53 @@ abstract class WebExtension(
...
      */
     abstract fun disconnectPort(name: String, session: EngineSession? = null)
+
+    /**
+     * Registers an [ActionHandler] for this web extension. The handler will
+     * be invoked whenever browser and page action defaults change. To listen
+     * for session-specific overrides see registerActionHandler(
+     * EngineSession, ActionHandler).
+     */
+    abstract fun registerActionHandler(actionHandler: ActionHandler)
```
Comment (psymoon) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt]: This appears to be registered for the lifecycle of an application. Is there a reason behind that?

File: components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt (hunk 7 above)
Comment (csadilek) [reply components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: 👍 let's save that line :)

Comment (csadilek) [reply components/support/webextensions/src/main/java/mozilla/components/support/webextensions/WebExtensionSupport.kt]: Updated.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/BrowserAction.kt (hunk 8 above)
Comment (csadilek) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/BrowserAction.kt]: This icon is specified by extension itself either in the manifest or via `setIcon`:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/browser_action
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/browserAction/setIcon

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt (hunk 9 above)
Comment (csadilek) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt]: So, it's registered globally on the webextension, yes. Later I think we'll remove it when extensions are disabled or uninstalled. We're not doing any UI work in these handlers, just making sure the store is up-to-date.

The UI components observing on the store (e.g. your `WebExtensionsToolbarFeature`) are all tied to a lifecycle (e.g. the BrowserFragment). So, we should be good here.

**Comment by csadilek** [comment]
bors r=psymoon,Amejia481
